cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/DefaultHttpServiceResponder.java

```diff
@@ -154,10 +152,20 @@ public void sendError(int status, String errorMessage) {
    * @param headers the headers to be sent back
    */
   @Override
-  public void send(int status, ByteBuffer content, String contentType, Multimap<String, String> headers) {
-    responder.sendContent(HttpResponseStatus.valueOf(status),
-                          ChannelBuffers.wrappedBuffer(content), contentType, headers);
-    emitMetrics(status);
+  public void send(final int status, ByteBuffer content, final String contentType, Multimap<String, String> headers) {
+    bufferedResponse = new BufferedResponse(status, content, contentType, headers);
```

Comment:
You can't send twice with the current logic. Any time you call one of the `send` methods, the `bufferedResponse` gets reset to the newest one. And since we call `execute` only once, only the last response set will be sent. This also helps since if the user sets a response, we can reset that response to something else on transaction failures.